### PR TITLE
Make measurements resubmission faster

### PR DESCRIPTION
**Pull request by bassosimone**
# Description

This diff takes advantage of recents changes in mk* libraries [1] to implement faster measurements resubmission [2] for iOS.

We vendor mkcurl v0.12.0, mkbouncer v0.2.0, and a development version of mkcollector [3], to have code able to reuse existing open HTTP connections.

We implement a new MKReporter class that uses new code in mkcollector such that we open and close reports only when the measurements we are submitting are clearly part of different reports. That is, the app does not need to concern itself with this problem.

We do not anticipate specific issues in integrating with the OONI app for iOS because resubmission is sync code and therefore there are no significant concerns around keeping objects alive.

# Scope of the review

Please, ignore the files in `./vendor`. They are copied from other repositories where there is very high code coverage and they have also already been reviewed. Focus on the newly added ObjectiveC code and specifically on:

- [x] correctness
- [x] whether the API seems appropriate
- [x] whether more tests are needed

# Roll out plan

- [x] I will tag a stable version of mkall-ios out of this branch. *This means that the branch will qualify as being merged without squashing*. The reason why this operation is safe is because the only change is this new `MKReporter` interface and all the rest is unchanged. The release message will clarify that this interface is still unstable and subject to changes.

- [x] Then, I will create a branch in the https://github.com/ooni/probe-ios repository using this new version and I will make sure that we can smoothly transition to the new API.

- [x] When https://github.com/measurement-kit/mkcollector/pull/25 is approved, I will tag a new stable release of mkcollector and vendor that into this repository.

- [x] Once we have updated the probe-ios codebase, I will remove the previous interface for submitting measurements because at this point that has become obsolete.

- [x] This will open up the opportunity of upgrading this repository to MK v0.10.5, where a number of vendored files that where used by resubmission have now been removed.

# Notes

1. measurement-kit/measurement-kit#1867

2. ooni/probe#861

3. measurement-kit/mkcollector#25

### CODE REVIEW

File: mkall/MKReporter.mm
Comment (bassosimone) [inline mkall/MKReporter.mm]: Here we also need to implement `deinit`.

File: mkallTests/MKReporterTests.mm
Comment (bassosimone) [inline mkallTests/MKReporterTests.mm]: This function can be written in a much more compact way

File: mkallTests/MKReporterTests.mm
Comment (bassosimone) [inline mkallTests/MKReporterTests.mm:68]: Maybe `makeSerializedMeasurement`?

File: mkallTests/MKReporterTests.mm
Comment (bassosimone) [inline mkallTests/MKReporterTests.mm]: The above interface should not be here

File: mkallTests/MKReporterTests.mm
Comment (bassosimone) [inline mkallTests/MKReporterTests.mm]: This needs to be addressed before moving forward.

**Review by bassosimone — COMMENTED** [review]
Pointed out places where more work is needed before others should review this code.

File: mkall/MKReporter.h
Comment (bassosimone) [inline mkall/MKReporter.h:18]: We should write tests to make sure this is updated.

File: mkall/MKReporter.h
Comment (bassosimone) [inline mkall/MKReporter.h:21]: We should write tests to make sure this is updated.

File: mkall/MKReporterInternal.h
Comment (bassosimone) [inline mkall/MKReporterInternal.h]: Because this is now internal, we can probably use the mkcollector.hpp struct directly

File: mkallTests/MKReporterTests.mm
Comment (bassosimone) [inline mkallTests/MKReporterTests.mm]: Add `// @interface ...`

File: mkallTests/MKReporterTests.mm
Comment (bassosimone) [inline mkallTests/MKReporterTests.mm]: Fix the docstring

**Review by bassosimone — COMMENTED** [review]
More small changes required before blessing this as ready for review by @lorenzoPrimi

File: mkall/MKReporter.mm
Comment (lorenzoPrimi) [inline mkall/MKReporter.mm]: Usually the conventional way to do this in Obj-C is
```
    self = [super init];
    if (self) {
//things
}
return self


```

**Review by lorenzoPrimi — APPROVED** [review]

File: mkall/MKReporter.mm
Comment (bassosimone) [reply mkall/MKReporter.mm]: Will change the code, thanks for letting me know!

File: mkall/MKReporter.h
Comment (bassosimone) [inline mkall/MKReporter.h]: Because the timeout is a per-request property, it's unfortunate to have it here. It would be much more convenient if the timeout was instead passed to `MKReporterTask:submit`.

**Review by bassosimone — COMMENTED** [review]
More changes to be implemented to make this work correctly for iOS

**Comment by bassosimone** [comment]
@lorenzoPrimi please take another look

**Review by lorenzoPrimi — APPROVED** [review]
Test it on the app as well

**Comment by bassosimone** [comment]
> Test it on the app as well

Sure, please see: https://github.com/ooni/probe-ios/pull/302

**Comment by bassosimone** [comment]
I'm going to update ooni/probe-ios#302. @lorenzoPrimi, please take another look.

File: mkallTests/MKReporterTests.mm
Comment (lorenzoPrimi) [inline mkallTests/MKReporterTests.mm:117]: shouldn't @ be only in front of string? I you want an int should be only
`@"bouncer_error": 0`, or am I missing something?

Comment (bassosimone) [reply mkallTests/MKReporterTests.mm:117]: The `@` in front of a number creates a `NSNumber`. I need to insert `NSNumbers` there mainly because that is what I do in the code that measures what happens.

Comment (lorenzoPrimi) [reply mkallTests/MKReporterTests.mm:117]: didn't know that, I always used `[NSNumber numberWithDouble:1000]`

**Comment by lorenzoPrimi** [comment]
Approved

**Pull request merged**
